week-4-tests-and-persistence/w04e02-test-nest/src/users/auth/auth.service.spec.ts

```diff
@@ -15,4 +17,20 @@ describe('AuthService', () => {
   it('should be defined', () => {
     expect(service).toBeDefined();
   });
+
+  it('should throw error when wrong credentials', async () => {
+    await expect(
+      service.signIn('jakub@example.pl', 'pass123'),
+    ).rejects.toThrowError();
+  });
+
+  it('should return token when correct credentials', async () => {
+    const token = await service.signIn('peter@myclient.com', 'som3pa55w0rd');
+    expect(token).toBeDefined();
+  });
+
+  it('should sign out', async () => {
```

Comment:
Tutaj musimy się zastanowić co to znaczy wylogować.

Jak wejdziesz w implementację faktyczną to `accessToken` będzie `null`.
Ale my zafejkowaliśmy `saveToken`
W tym układzie, potrzebujemy szpiega, który sprawdzi, czy `saveToken` wywołuje się jako "naszeIdUsera + `null`"

Żeby to uczynić musimy mieć wcześniej dostęp do `usersService` jako instancji.
Więc - potrzebujemy wyciągnąć go z `TestingModule` w `beforeEach`

na górze:
```typescript
 let userService: UsersService;
```

w `beforeEach` (na końcu):
```typescript
userService = module.get(UsersService);
```

I teraz zanawa w szpiega w teście:

```typescript
it('should sign out, making accessToken null', async () => {
    const spyTheNullify = jest.spyOn(userService, 'saveToken');
    const userId = 29;

    await service.signOut(userId);

    expect(spyTheNullify).toHaveBeenCalledWith(userId, null);
  });
```